Readability.js

```diff
@@ -1316,6 +1316,89 @@ Readability.prototype = {
+   * whether as a direct child or as its descendants.
+   *
+   * @param Element
+  **/
+  _isSingleImage: function(node) {
+    if (node.tagName === "IMG") {
+      return true;
+    }
+
+    if (node.children.length !== 1 || node.textContent.trim() !== "") {
+      return false;
+    }
+
+    return this._isSingleImage(node.children[0]);
+  },
+
+  /**
+   * Find all <noscript> that are located after <img> nodes, and which contain only one
+   * <img> element. Replace the first image with the image from inside the <noscript> tag,
+   * and remove the <noscript> tag. This improves the quality of the images we use on
+   * some sites (e.g. Medium).
+   *
+   * @param Element
+  **/
+  _unwrapNoscriptImages: function(doc) {
+    // Find img without source and remove it. This is done to prevent a placeholder
+    // img is replaced by img from noscript in next step.
+    var imgs = Array.from(doc.getElementsByTagName("img"));
+    this._forEachNode(imgs, function(img) {
+      var src = img.getAttribute("src");
+      var srcset = img.getAttribute("srcset");
+      var dataSrc = img.getAttribute("data-src");
+      var dataSrcset = img.getAttribute("data-srcset");
+
+      if (!src && !srcset && !dataSrc && !dataSrcset) {
+        img.parentNode.removeChild(img);
+      }
+    });
+
+    // Next find noscript and try to extract its image
+    var noscripts = Array.from(doc.getElementsByTagName("noscript"));
+    this._forEachNode(noscripts, function(noscript) {
+      // Parse content of noscript and make sure it only contains image
+      var tmp = doc.createElement("div");
+      tmp.innerHTML = noscript.innerHTML;
+      if (!this._isSingleImage(tmp)) {
+        return;
+      }
+
+      // If noscript has previous sibling and it only contains image,
+      // replace it with noscript content. However, in some case there
+      // is possibility that img inside noscript has lower quality
+      // than the one in previous sibling, so we will keep the src
+      // and srcset attribute from old img as data attribute for
+      // img from noscript.
+      var prevElement = noscript.previousElementSibling;
+      if (prevElement && this._isSingleImage(prevElement)) {
+        var prevImg = prevElement;
+        if (prevImg.tagName !== "IMG") {
+          prevImg = prevElement.getElementsByTagName("img")[0];
+        }
+
+        var newImg = tmp.getElementsByTagName("img")[0];
+        var newImgSrc = newImg.getAttribute("src");
+        var newImgSrcset = newImg.getAttribute("srcset");
+        var prevImgSrc = prevImg.getAttribute("src");
+        var prevImgSrcset = prevImg.getAttribute("srcset");
+
+        if (prevImgSrc && prevImgSrc !== newImgSrc) {
+          newImg.setAttribute("data-old-src", prevImgSrc);
+        }
+
+        if (prevImgSrcset && prevImgSrcset !== newImgSrcset) {
+          newImg.setAttribute("data-old-srcset", prevImgSrcset);
+        }
+
+        noscript.parentNode.replaceChild(tmp.children[0], prevElement);
```

Comment:
Can we add code along the lines of:

```js
for (var i = 0; i < prevImg.attributes.length; i++) {
  var attr = prevImg.attributes[i];
  if (attr.name == "src" || attr.name == "srcset" || newImg.hasAttribute(attr.name)) {
    continue;
  }
  if (/\.(jpg|jpeg|png|webp)/i).test(attr.value)) {
    newImg.setAttribute(attr.name, attr.value);
  }
}
```

here so we preserve image-related attributes?

Reply:
I've added it in [be67f18](https://github.com/mozilla/readability/pull/588/commits/be67f18f88c08501372c809032e768f9092683db)